### PR TITLE
Async storage set, get, and remove Item implementation

### PR DESCRIPTION
**Pull request by Dileepnidi**
- Async storage set,get and remove Item implementation
- added task scheduler.

### CODE REVIEW

File: skia/BUILD.gn [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 import("//react-native/common.gni")
 
-config("LibRnAsyncStorage_config") {
+config("LibRNSAsyncStorage_config") {
```
Comment (Dileepnidi) [inline skia/BUILD.gn]: change name to LibRSAsyncStorage_config

File: skia/RSAsyncStorageModule.cpp [hunk 2]
```diff
@@ -9,11 +9,55 @@
...
+        int length = outfile_.tellg();
+        outfile_.seekg (0, outfile_.beg);
+        char * buffer = new char [length];
+        // read data as a block:
+        outfile_.read (buffer,length);
+        RNS_LOG_ERROR(buffer);
+        string Str(buffer);
+        try {
+          globalDynamic_ = parseJson(Str);
+          totalSize = length;
```
Comment (Dileepnidi) [inline skia/RSAsyncStorageModule.cpp]: add length condition if its zero

File: skia/RSAsyncStorageModule.cpp [hunk 3]
```diff
@@ -9,11 +9,55 @@
...
+      }else {
+        RNS_LOG_ERROR("Cannot open file");
+        return;
+      }
+
+    });
+}
+
+RSAsyncStorageModule::~RSAsyncStorageModule() {
+  outfile_.close();
```
Comment (Dileepnidi) [inline skia/RSAsyncStorageModule.cpp]: check outfile is valid or not, then close.

File: skia/RSAsyncStorageModule.h [hunk 4]
```diff
@@ -9,14 +9,32 @@
...
 class RSAsyncStorageModule : public module::CxxModule {
+private:
+  void multiGet(dynamic args, CxxModule::Callback cb);
+  void multiSet(dynamic args, CxxModule::Callback cb);
+  void multiRemove(dynamic args, CxxModule::Callback cb);
+  void mergeItem(dynamic args, CxxModule::Callback cb);
+  void getAllKeys(dynamic args, CxxModule::Callback cb);
+  void clear(dynamic args, CxxModule::Callback cb);
+  void asyncWorkerThread();
+  dynamic globalDynamic_ = dynamic::object;
```
Comment (Dileepnidi) [inline skia/RSAsyncStorageModule.h]: applocalDataFile

File: skia/RSAsyncStorageModule.h [hunk 5]
```diff
@@ -9,14 +9,32 @@
...
+  fstream outfile_;
```
Comment (Dileepnidi) [inline skia/RSAsyncStorageModule.h]: applocalFile

File: skia/RSAsyncStorageModule.cpp [hunk 6]
```diff
@@ -28,49 +72,123 @@ auto RSAsyncStorageModule::getMethods() -> std::vector<Method> {
...
+    RNS_LOG_ERROR("totalSize" << totalSize);
+    if(isWriteScheduled_)
+      return;
+    RNS_LOG_DEBUG("Scheduling to write the file");
+    taskRunner_->scheduleDispatch( [&](){
+      RNS_LOG_DEBUG ("writing to json file");
+      string str = toJson(globalDynamic_);
+      if(!outfile_.is_open()) {
+        dynamic errors = folly::dynamic::object();
+        errors["message"] = "not able to set Item operation";
```
Comment (Dileepnidi) [inline skia/RSAsyncStorageModule.cpp]: failed to write the data

File: skia/RSAsyncStorageModule.cpp [hunk 7]
```diff
@@ -28,49 +72,123 @@ auto RSAsyncStorageModule::getMethods() -> std::vector<Method> {
...
+        return;
+      }
+      RNS_LOG_ERROR("SIZE OF file" << str.length());
+      // Total storage size is capped at 6 MB by default.
+      if(totalSize < ASYNC_DEFAULT_MAX_CACHE_LIMIT) {
+        filesystem::resize_file("SimpleViewApp.json", 0);
+        outfile_.seekp(0);
+        outfile_ << str;
+        outfile_.flush();
+        isWriteScheduled_ = false;
```
Comment (Dileepnidi) [inline skia/RSAsyncStorageModule.cpp]: out the if condition

File: skia/RSAsyncStorageModule.cpp [hunk 8]
```diff
@@ -28,49 +72,123 @@ auto RSAsyncStorageModule::getMethods() -> std::vector<Method> {
...
+      if(!outfile_.is_open()) {
+        dynamic errors = folly::dynamic::object();
+        errors["message"] = "not able to set Item operation";
+        errors["key"] = args[0][0].asString();
+        cb({folly::dynamic::array(errors)});
+        return;
+      }
+      RNS_LOG_ERROR("SIZE OF file" << str.length());
+      // Total storage size is capped at 6 MB by default.
+      if(totalSize < ASYNC_DEFAULT_MAX_CACHE_LIMIT) {
```
Comment (Dileepnidi) [inline skia/RSAsyncStorageModule.cpp]: totalSize checked can be added before the scheduleDispatch

File: skia/RSAsyncStorageModule.cpp [hunk 9]
```diff
@@ -28,49 +72,123 @@ auto RSAsyncStorageModule::getMethods() -> std::vector<Method> {
...
+    }
+    totalSize += requiredLength;
+    RNS_LOG_ERROR("totalSize" << totalSize);
+    if(isWriteScheduled_)
+      return;
+    RNS_LOG_DEBUG("Scheduling to write the file");
+    taskRunner_->scheduleDispatch( [&](){
+      RNS_LOG_DEBUG ("writing to json file");
+      string str = toJson(globalDynamic_);
+      if(!outfile_.is_open()) {
```
Comment (Dileepnidi) [inline skia/RSAsyncStorageModule.cpp]: add this check in the starting of this method

File: skia/BUILD.gn [hunk 10]
```diff
@@ -27,6 +27,7 @@ rn_cxx_component("RNSAsyncStorage") {
   public_configs += [ "//build/secondary/folly:folly_config" ]
   deps = [
     "//folly",
```
Comment (Dileepnidi) [inline skia/BUILD.gn]: compile and check folly is required or not

File: skia/RSAsyncStorageModule.cpp [hunk 11]
```diff
@@ -9,11 +9,54 @@
 
 #include "ReactSkia/utils/RnsLog.h"
 #include "RSAsyncStorageModule.h"
+#include <folly/json.h>
```
Comment (Dileepnidi) [inline skia/RSAsyncStorageModule.cpp]: add in the starting

File: skia/RSAsyncStorageModule.cpp [hunk 12]
```diff
@@ -9,11 +9,54 @@
...
 
-RSAsyncStorageModule::RSAsyncStorageModule(){
+RSAsyncStorageModule::RSAsyncStorageModule() {
+  taskRunner_ = std::make_unique<RnsShell::TaskLoop>();
+  workerThread_=std::thread (&RSAsyncStorageModule::asyncWorkerThread,this);
+  taskRunner_->waitUntilRunning();
+
+  taskRunner_->dispatch( [this]() {
+    appLocalFile_.open(FILE_PATH, ios::out | ios::in | ios::app);
+    if (appLocalFile_) {
```
Comment (munezbn) [inline skia/RSAsyncStorageModule.cpp]: This condition need to be replaced with fstream api check

File: skia/RSAsyncStorageModule.cpp [hunk 13]
```diff
@@ -9,11 +9,54 @@
...
+    }else {
+      RNS_LOG_ERROR("Cannot open file");
+      return;
+    }
+
+  });
+}
+
+RSAsyncStorageModule::~RSAsyncStorageModule() {
+  if(appLocalFile_) {
```
Comment (munezbn) [inline skia/RSAsyncStorageModule.cpp]: Clear appLocalDataFile_

File: skia/RSAsyncStorageModule.cpp [hunk 14]
```diff
@@ -28,49 +71,162 @@ auto RSAsyncStorageModule::getMethods() -> std::vector<Method> {
...
 
   };
 }
 
+
+void RSAsyncStorageModule::multiGet(dynamic args, CxxModule::Callback cb) {
+  taskRunner_->dispatch( [this,args,cb](){
+    dynamic errors = folly::dynamic::array;
+    dynamic resultArray = folly::dynamic::array;
+    for (auto& itemArray : args[0]) {
```
Comment (munezbn) [inline skia/RSAsyncStorageModule.cpp]: itemArray -> keystring

File: skia/RSAsyncStorageModule.cpp [hunk 15]
```diff
@@ -28,49 +71,162 @@ auto RSAsyncStorageModule::getMethods() -> std::vector<Method> {
...
 }
 
+
+void RSAsyncStorageModule::multiGet(dynamic args, CxxModule::Callback cb) {
+  taskRunner_->dispatch( [this,args,cb](){
+    dynamic errors = folly::dynamic::array;
+    dynamic resultArray = folly::dynamic::array;
+    for (auto& itemArray : args[0]) { 
+      auto pos = appLocalDataFile_.find(itemArray);
+      dynamic errorItem = folly::dynamic::object();
```
Comment (munezbn) [inline skia/RSAsyncStorageModule.cpp]: Should be moved inside if

File: skia/RSAsyncStorageModule.cpp [hunk 16]
```diff
@@ -28,49 +71,162 @@ auto RSAsyncStorageModule::getMethods() -> std::vector<Method> {
...
+      if(pos == appLocalDataFile_.items().end()){
+        errorItem["message"] = "Failed to found the get Item";
+        errorItem["key"] = itemArray;
+        errors.push_back(errorItem);
+      }else{
+        dynamic result = folly::dynamic::array;
+        result.push_back(itemArray);
+        result.push_back(appLocalDataFile_[itemArray]);
```
Comment (munezbn) [inline skia/RSAsyncStorageModule.cpp]: You need to push  pos-> first ( for key) and pos->second ( for value)

File: skia/RSAsyncStorageModule.cpp [hunk 17]
```diff
@@ -28,49 +71,162 @@ auto RSAsyncStorageModule::getMethods() -> std::vector<Method> {
...
+  });  
+}
+
+void RSAsyncStorageModule::multiSet(dynamic args, CxxModule::Callback cb) {
+  if(!appLocalFile_.is_open()) {
+    dynamic errors = folly::dynamic::object();
+    errors["message"] = "Failed to write the data";
+    errors["key"] = args[0][0].asString();
+    cb({folly::dynamic::array(errors)});
+    return;
```
Comment (munezbn) [inline skia/RSAsyncStorageModule.cpp]: This error needs to be generated for every key and should be inside the dispatch.

File: skia/RSAsyncStorageModule.cpp [hunk 18]
```diff
@@ -28,49 +71,162 @@ auto RSAsyncStorageModule::getMethods() -> std::vector<Method> {
...
+  }
+  taskRunner_->dispatch( [this,args,cb]() {
+    int requiredLength = 0;
+    for (auto& itemArray : args[0]) {
+      int oldLength = 0;
+      int newLength = (itemArray[0].asString()).length()+itemArray[1].asString().length()+6;
```
Comment (munezbn) [inline skia/RSAsyncStorageModule.cpp]: use EXTRA_CHARACTER_PADDING_LENGTH macro for constant 6 
// padding for commas and colons

File: skia/RSAsyncStorageModule.cpp [hunk 19]
```diff
@@ -28,49 +71,162 @@ auto RSAsyncStorageModule::getMethods() -> std::vector<Method> {
...
+      // value size cannot be larger than 2 MB
+      if((itemArray[1].asString()).length() < ASYNC_VALUE_DEFAULT_MAX_CACHE_LIMIT ) {
```
Comment (munezbn) [inline skia/RSAsyncStorageModule.cpp]: ASYNC_STORAGE_VALUE_MAX_LENGTH

File: skia/RSAsyncStorageModule.cpp [hunk 20]
```diff
@@ -28,49 +71,162 @@ auto RSAsyncStorageModule::getMethods() -> std::vector<Method> {
...
+          oldLength = (itemArray[0].asString()).length()+(appLocalDataFile_[itemArray[0]].asString()).length()+6;
+          if(oldLength < newLength) {
+            requiredLength = newLength - oldLength;
+          }else{
+            requiredLength = oldLength-newLength;
+          }
+        }else{
+          requiredLength= newLength;
+        }
+        appLocalDataFile_[itemArray[0].asString()] = itemArray[1].asString();
```
Comment (munezbn) [inline skia/RSAsyncStorageModule.cpp:166]: You should not storage key here. It should be after max size check.

File: skia/RSAsyncStorageModule.cpp [hunk 21]
```diff
@@ -28,49 +71,162 @@ auto RSAsyncStorageModule::getMethods() -> std::vector<Method> {
...
+      }
+    }
+    totalSize_ += requiredLength;
```
Comment (munezbn) [inline skia/RSAsyncStorageModule.cpp]: All this should be done only if 
(current totalsize_ + requiredLength ) <  ASYNC_STORAGE_FILE_MAX_LENGTH

File: skia/RSAsyncStorageModule.cpp [hunk 22]
```diff
@@ -28,49 +71,162 @@ auto RSAsyncStorageModule::getMethods() -> std::vector<Method> {
...
+    if(isWriteScheduled_)
+      return;
```
Comment (munezbn) [inline skia/RSAsyncStorageModule.cpp]: All these should be done for every key. Should be inside for loop.
Need to schedule only one file write after for loop ( you can use a local bool flag )
set global flag once we schedule

File: skia/RSAsyncStorageModule.cpp [hunk 23]
```diff
@@ -28,49 +71,162 @@ auto RSAsyncStorageModule::getMethods() -> std::vector<Method> {
...
+    for (auto& itemArray : args[0]) {
+      auto pos = appLocalDataFile_.find(itemArray);
+      if(pos == appLocalDataFile_.items().end()) {
+        errors["message"] = "Failed to remove the data";
+        errors["key"] = args[0][0].asString();
+      }else {
+        appLocalDataFile_.erase(itemArray);
+      }
+    }
+    taskRunner_->scheduleDispatch( [&](){
```
Comment (munezbn) [inline skia/RSAsyncStorageModule.cpp]: Need to be scheduled based on global isScheduled check

File: skia/RSAsyncStorageModule.cpp [hunk 24]
```diff
@@ -28,49 +71,162 @@ auto RSAsyncStorageModule::getMethods() -> std::vector<Method> {
...
+});
+
+}
+
+void RSAsyncStorageModule::writeToFile(){
+  string str = toJson(appLocalDataFile_);
+  filesystem::resize_file(FILE_PATH, 0);
+  appLocalFile_.seekp(0);
+  appLocalFile_ << str;
+  appLocalFile_.flush();
```
Comment (munezbn) [inline skia/RSAsyncStorageModule.cpp:193]: set  isWriteScheduled_ = false;  inside this file

File: skia/RSAsyncStorageModule.h [hunk 25]
```diff
@@ -5,18 +5,44 @@
...
+#include "rns_shell/platform/linux/TaskLoop.h"
 #include <cxxreact/CxxModule.h>
-
 #include "ReactSkia/utils/RnsUtils.h"
 
+#ifndef FILE_PATH
+#define FILE_PATH "SimpleViewApp.json"
+#endif
+
+#define ASYNC_SRG_TIMEOUT 5000
```
Comment (munezbn) [inline skia/RSAsyncStorageModule.h]: ASYNC_STORGAE_FILE_WRITE_TIMEOUT

File: skia/RSAsyncStorageModule.h [hunk 26]
```diff
@@ -5,18 +5,44 @@
...
+#define ASYNC_STORAGE_DEFAULT_MAX_CACHE_LIMIT 6*1024*1024 //6,971,520 bytes
+#define ASYNC_VALUE_DEFAULT_MAX_CACHE_LIMIT 2*1024*1024 //2,971,520 bytes
```
Comment (munezbn) [inline skia/RSAsyncStorageModule.h]: enclose with braces

File: skia/RSAsyncStorageModule.cpp [hunk 27]
```diff
@@ -28,49 +69,181 @@ auto RSAsyncStorageModule::getMethods() -> std::vector<Method> {
...
+        errorItem["key"] = keyString;
+        errors.push_back(errorItem);
+      }else{
+        dynamic result = folly::dynamic::array;
+        result.push_back(pos->first);
+        result.push_back(pos->second);
+        resultArray.push_back(result);
+      }
+    }
+    cb({(errors.empty() ? nullptr : errors), resultArray});
```
Comment (munezbn) [inline skia/RSAsyncStorageModule.cpp]: Can we check if same check is required for empty result array ?

File: skia/RSAsyncStorageModule.cpp [hunk 28]
```diff
@@ -28,49 +69,181 @@ auto RSAsyncStorageModule::getMethods() -> std::vector<Method> {
...
+        dynamic errorItem = folly::dynamic::object();
+        errorItem["message"] = "Failed to write the data";
+        errorItem["key"] = itemArray[0].asString();
+        errors.push_back(errorItem);
+      }
+      cb({folly::dynamic::array(errors)});
+      return;
+    }
+    int requiredLength = 0;
+    bool needsFileWrite;
```
Comment (munezbn) [inline skia/RSAsyncStorageModule.cpp]: initialize with false;

File: skia/RSAsyncStorageModule.cpp [hunk 29]
```diff
@@ -28,49 +69,181 @@ auto RSAsyncStorageModule::getMethods() -> std::vector<Method> {
...
+            requiredLength = newLength - oldLength;
+          }else{
+            requiredLength = oldLength-newLength;
+          }
+        }else{
+          requiredLength= newLength;
+        }
+      } else {
+         dynamic errorItem = folly::dynamic::object();
+         errorItem["message"] = "Failed to write the data, value size is larger than 2 MB";
```
Comment (munezbn) [inline skia/RSAsyncStorageModule.cpp]: Can we use ASYNC_STORAGE_VALUE_MAX_LENGTH to print the value 2

File: skia/RSAsyncStorageModule.cpp [hunk 30]
```diff
@@ -28,49 +69,181 @@ auto RSAsyncStorageModule::getMethods() -> std::vector<Method> {
...
+         errorItem["key"] = itemArray[0].asString();
+         errorList.push_back(errorItem);
+         continue;
+      }
+      if(( totalSize_ + requiredLength) < ASYNC_STORAGE_FILE_MAX_LENGTH){
+        appLocalDataFile_[itemArray[0].asString()] = itemArray[1].asString();
+        needsFileWrite = true;
+        totalSize_ += requiredLength;
+      }else{
+        RNS_LOG_WARN("async storage size is greater than 6mb");
```
Comment (munezbn) [inline skia/RSAsyncStorageModule.cpp]: Can we use ASYNC_STORAGE_FILE_MAX_LENGTH to print value 6

File: skia/RSAsyncStorageModule.cpp [hunk 31]
```diff
@@ -28,49 +69,181 @@ auto RSAsyncStorageModule::getMethods() -> std::vector<Method> {
...
+void RSAsyncStorageModule::multiRemove(dynamic args, CxxModule::Callback cb) {
+  taskRunner_->dispatch( [this,args,cb](){
+    if(!appLocalFile_.is_open()) {
+      dynamic errors = folly::dynamic::array;
+      for(auto& keyString : args[0]){
+        dynamic errorItem = folly::dynamic::object();
+        errorItem["message"] = "Failed to remove the data";
+        errorItem["key"] = keyString;
+        errors.push_back(errorItem);
+      }
```
Comment (munezbn) [inline skia/RSAsyncStorageModule.cpp:206]: I think we can make a private function to do this. t is done at two places

**Review by munezbn — APPROVED** [review]

**Pull request merged**